Torch2 compatibility

RaulPPelaez opened the pull request.

raimis commented:
This failed because there is no NNPOps for PyTorch 2.0. Should be solved by https://github.com/conda-forge/nnpops-feedstock/pull/22

raimis commented:
NNPOs has been rebuilt for PyTorch 2.0 (https://github.com/conda-forge/nnpops-feedstock/pull/22). Triggered a rerun of the tests.

RaulPPelaez commented:
Looking into it

RaulPPelaez commented:
The C++ tests pass without issue (even the CUDA/OCL ones in my machine).
Every python tests segfaults in the CI and also in my setup, though. I do not know why -.-

RaulPPelaez commented:
I now know why, this line is the culprit:
https://github.com/openmm/openmm-torch/blob/b76deb4df8661713925a6254613823df8601c9dc/python/openmmtorch.i#L47
Used to work, it does not work anymore... Lets see if there is an alternative.

RaulPPelaez commented:
The actual error goes like this:
```bash
PyObject_IsInstance (inst=0x7ffff72d4b30, cls=0x0) at /usr/local/src/conda/python-3.9.16/Objects/abstract.c:2557                                                             
2557    /usr/local/src/conda/python-3.9.16/Objects/abstract.c: No such file or directory.                                                                                    
(gdb) bt                                                                                                                                                                     
#0  PyObject_IsInstance (inst=0x7ffff72d4b30, cls=0x0) at /usr/local/src/conda/python-3.9.16/Objects/abstract.c:2557                                                         
#1  0x00007ffff72a6954 in torch::jit::as_module(pybind11::handle) ()                                                                                                         
   from /shared/raul/mambaforge/envs/openmmtorch2/lib/python3.9/site-packages/_openmmtorch.cpython-39-x86_64-linux-gnu.so                                                    
#2  0x00007ffff72a909a in _wrap_new_TorchForce () from /shared/raul/mambaforge/envs/openmmtorch2/lib/python3.9/site-packages/_openmmtorch.cpython-39-x86_64-linux-gnu.so     
#3  0x000055555569ddf0 in cfunction_call (func=0x7ffff72ebe00, args=<optimized out>, kwargs=<optimized out>)                                                                 
    at /usr/local/src/conda/python-3.9.16/Objects/methodobject.c:552
```
Which looks similar to whatever happened here:
https://github.com/openmm/openmm/issues/2586

raimis commented:
It seems that [PyObject_IsInstance](https://docs.python.org/3/c-api/object.html#c.PyObject_IsInstance) receives `NULL` as its second argument.

RaulPPelaez commented:
I think I got it, I will lay out my findings in case someone has some insight.
This is the definition of as_module in torch:
```c++

inline c10::optional<Module> as_module(py::handle obj) {
  static py::handle ScriptModule =
      py::module::import("torch.jit").attr("ScriptModule");
  if (py::isinstance(obj, ScriptModule)) {
    return py::cast<Module>(obj.attr("_c"));
  }
  return c10::nullopt;
}
```
I modified the swig typecheck to understand whats going on:
```c++
%typecheck(SWIG_TYPECHECK_POINTER) const torch::jit::Module& {
  std::cerr<<"Checking if object is a torch module"<<std::endl;
    py::object o = py::reinterpret_borrow<py::object>($input);
    std::cerr<<"Input: "<<$input<<std::endl;
    std::cerr<<"Object: "<<o<<std::endl;
    py::handle ScriptModule = py::module::import("torch.jit").attr("ScriptModule");
    std::cerr<<ScriptModule<<std::endl;
    std::cerr<<"Is a torch module: "<<py::isinstance(o, ScriptModule)<<std::endl;
    std::cerr<<"Using object: "<<$1<<std::endl;
    $1 = torch::jit::as_module(o).has_value() ? 1 : 0;
    std::cerr<<"Is a torch module: "<<$1<<std::endl;
}
```

I tested by simply creating a TorchForce object with an existing module.pt file:
```python

class SimpleMod(torch.nn.Module):
    def forward(self, positions):
        return (torch.sum(positions**2), -2.0 * positions)

module = torch.jit.script(SimpleMod())
module.save("module.pt")
torch_force = ot.TorchForce("module.pt")
```
The code crashes but prints this:

```shell
Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
Testing with module <class '__main__.SimpleMod'>
RecursiveScriptModule(original_name=SimpleMod)
Checking if object is a torch module
Input: 0x7f25eafc6430
Object: module.pt
<class 'torch.jit._script.ScriptModule'>
Is a torch module: 0
Using object: 0
Segmentation fault (core dumped)
```
Funny enough, in my modified typecheck I am doing basically the same as as_module without crashing. There is a small difference, though, the as_module has the import as "static" while my typecheck does not.

Indeed, removing the "static" in as_module fixes the issue and the tests pass in my machine.
Since I obviously cannot modify as_module I changed the typecheck to:
```c++
%typecheck(SWIG_TYPECHECK_POINTER) const torch::jit::Module& {
    py::object o = py::reinterpret_borrow<py::object>($input);
    py::handle ScriptModule = py::module::import("torch.jit").attr("ScriptModule");
    $1 = py::isinstance(o, ScriptModule);
}
```
Lets see if the CI likes it as much as my computer.

RaulPPelaez commented:
Success!
The only truly supported way for a Module to cross the python/C++ barrier is via serialization/deserialization.
I am not sure what the intended use of torch::jit::as_module is, but it is not documented (nor commented, for that matter). I modified the swig maps to serialize/deserialize and now all tests pass.

RaulPPelaez commented:
All tests (CUDA included) pass in my machine. This is ready for review.

RaulPPelaez commented:
I am getting this error:


<details>
  <summary>Click me</summary>

```shell

TestInteroperability.py ....FF..                                                                                                                                      [100%]

================================================================================= FAILURES ==================================================================================
__________________________________________________________________________ testTorchANI[CUDA-True] __________________________________________________________________________

use_cv_force = True, platform = <openmm.openmm.Platform; proxy of <Swig Object of type 'OpenMM::Platform *' at 0x7f8b925ae0f0> >

    @pytest.mark.skipif(platform.system() == 'Darwin', reason='There is no NNPOps package for MacOS')
    @pytest.mark.parametrize('use_cv_force', [True, False])
    @pytest.mark.parametrize('platform', ['Reference', 'CPU', 'CUDA', 'OpenCL'])
    def testTorchANI(use_cv_force, platform):

        if pt.cuda.device_count() < 1 and platform == 'CUDA':
            pytest.skip('A CUDA device is not available')

        import NNPOps # There is no NNPOps package for MacOS
        import torchani

        class Model(pt.nn.Module):

            def __init__(self):
                super().__init__()
                self.register_buffer('atomic_numbers', pt.tensor([[1, 1]]))
                self.model = torchani.models.ANI2x(periodic_table_index=True)
                self.model = NNPOps.OptimizedTorchANI(self.model, self.atomic_numbers)

            def forward(self, positions):
                positions = positions.float().unsqueeze(0) * 10 # nm --> Ang
                return self.model((self.atomic_numbers, positions)).energies[0] * 2625.5 # Hartree --> kJ/mol

        # Create a system
        system = mm.System()
        for _ in range(2):
            system.addParticle(1.0)
        positions = pt.tensor([[-5, 0.0, 0.0], [5, 0.0, 0.0]], requires_grad=True)

        with NamedTemporaryFile() as model_file:

            # Save the model
            pt.jit.script(Model()).save(model_file.name)

            # Compute reference energy and forces
            model = pt.jit.load(model_file)
            ref_energy = model(positions)
            ref_energy.backward()
            ref_forces = positions.grad

            # Create a force
            force = ot.TorchForce(model_file.name)
            if use_cv_force:
                # Wrap TorchForce into CustomCVForce
                cv_force = mm.CustomCVForce('force')
                cv_force.addCollectiveVariable('force', force)
                system.addForce(cv_force)
            else:
                system.addForce(force)

            # Compute energy and forces
            integ = mm.VerletIntegrator(1.0)
            platform = mm.Platform.getPlatformByName(platform)
            context = mm.Context(system, integ, platform)
            context.setPositions(positions.detach().numpy())
>           state = context.getState(getEnergy=True, getForces=True)

TestInteroperability.py:65:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <openmm.openmm.Context; proxy of <Swig Object of type 'OpenMM::Context *' at 0x7f8b925ae1b0> >, getPositions = False, getVelocities = False, getForces = True
getEnergy = True, getParameters = False, getParameterDerivatives = False, getIntegratorParameters = False, enforcePeriodicBox = False, groups = -1

    def getState(self, getPositions=False, getVelocities=False,
                 getForces=False, getEnergy=False, getParameters=False,
                 getParameterDerivatives=False, getIntegratorParameters=False,
                 enforcePeriodicBox=False, groups=-1):
        """Get a State object recording the current state information stored in this context.

        Parameters
        ----------
        getPositions : bool=False
            whether to store particle positions in the State
        getVelocities : bool=False
            whether to store particle velocities in the State
        getForces : bool=False
            whether to store the forces acting on particles in the State
        getEnergy : bool=False
            whether to store potential and kinetic energy in the State
        getParameters : bool=False
            whether to store context parameters in the State
        getParameterDerivatives : bool=False
            whether to store parameter derivatives in the State
        getIntegratorParameters : bool=False
            whether to store integrator parameters in the State
        enforcePeriodicBox : bool=False
            if false, the position of each particle will be whatever position
            is stored in the Context, regardless of periodic boundary conditions.
            If true, particle positions will be translated so the center of
            every molecule lies in the same periodic box.
        groups : set={0,1,2,...,31}
            a set of indices for which force groups to include when computing
            forces and energies. The default value includes all groups. groups
            can also be passed as an unsigned integer interpreted as a bitmask,
            in which case group i will be included if (groups&(1<<i)) != 0.
        """
        try:
    # is the input integer-like?
            groups_mask = int(groups)
        except TypeError:
            if isinstance(groups, set):
    # nope, okay, then it should be an set
                groups_mask = functools.reduce(operator.or_,
                        ((1<<x) & 0xffffffff for x in groups))
            else:
                raise TypeError('%s is neither an int nor set' % groups)
        if groups_mask >= 0x80000000:
            groups_mask -= 0x100000000
        types = 0
        if getPositions:
            types += State.Positions
        if getVelocities:
            types += State.Velocities
        if getForces:
            types += State.Forces
        if getEnergy:
            types += State.Energy
        if getParameters:
            types += State.Parameters
        if getParameterDerivatives:
            types += State.ParameterDerivatives
        if getIntegratorParameters:
            types += State.IntegratorParameters
>       state = _openmm.Context_getState(self, types, enforcePeriodicBox, groups_mask)
E       openmm.OpenMMException: The following operation failed in the TorchScript interpreter.
E       Traceback of TorchScript, serialized code (most recent call last):
E         File "code/__torch__/TestInteroperability/___torch_mangle_17.py", line 14, in forward
E           model = self.model
E           atomic_numbers = self.atomic_numbers
E           _1 = (model).forward((atomic_numbers, positions0), None, None, )
E                 ~~~~~~~~~~~~~~ <--- HERE
E           _2 = torch.mul(torch.select((_1)[1], 0, 0), 2625.5)
E           return _2
E         File "code/__torch__/NNPOps/OptimizedTorchANI.py", line 17, in forward
E           species_coordinates0 = (species_converter).forward(species_coordinates, None, None, )
E           aev_computer = self.aev_computer
E           species_aevs = (aev_computer).forward(species_coordinates0, cell, pbc, )
E                           ~~~~~~~~~~~~~~~~~~~~~ <--- HERE
E           neural_networks = self.neural_networks
E           species_energies = (neural_networks).forward(species_aevs, )
E         File "code/__torch__/NNPOps/SymmetryFunctions.py", line 39, in forward
E           _4 = ops.NNPOpsANISymmetryFunctions.operation(holder, torch.select(positions, 0, 0), cell0)
E           radial, angular, = _4
E           features = torch.unsqueeze(torch.cat([radial, angular], 1), 0)
E                                      ~~~~~~~~~ <--- HERE
E           return (species, features)
E
E       Traceback of TorchScript, original code (most recent call last):
E         File "/shared/raul/openmm-torch/python/tests/TestInteroperability.py", line 31, in forward
E               def forward(self, positions):
E                   positions = positions.float().unsqueeze(0) * 10 # nm --> Ang
E                   return self.model((self.atomic_numbers, positions)).energies[0] * 2625.5 # Hartree --> kJ/mol
E                          ~~~~~~~~~~ <--- HERE
E         File "/shared/raul/mambaforge/envs/openmmtorch2/lib/python3.9/site-packages/NNPOps/OptimizedTorchANI.py", line 52, in forward
E
E               species_coordinates = self.species_converter(species_coordinates)
E               species_aevs = self.aev_computer(species_coordinates, cell=cell, pbc=pbc)
E                              ~~~~~~~~~~~~~~~~~ <--- HERE
E               species_energies = self.neural_networks(species_aevs)
E               species_energies = self.energy_shifter(species_energies)
E         File "/shared/raul/mambaforge/envs/openmmtorch2/lib/python3.9/site-packages/NNPOps/SymmetryFunctions.py", line 121, in forward
E
E               radial, angular = operation(self.holder, positions[0], cell)
E               features = torch.cat((radial, angular), dim=1).unsqueeze(0)
E                          ~~~~~~~~~ <--- HERE
E
E               return species, features
E       RuntimeError: CUDA error: no kernel image is available for execution on the device
E       CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
E       For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
E       Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

../../../mambaforge/envs/openmmtorch2/lib/python3.9/site-packages/openmm/openmm.py:10009: OpenMMException
```
</details>

peastman commented:
```
E       CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
E       For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

Like it says!  If you set `CUDA_LAUNCH_BLOCKING=1`, you should get an error message that tells what the actual problem is.

peastman commented on python/openmmtorch.i:
```diff
@@ -28,23 +29,40 @@
     py::object o = py::reinterpret_borrow<py::object>($input);
-    module = torch::jit::as_module(o).value();
-    $1 = &module;
+    auto fileName = std::tmpnam(nullptr);
```
There's probably a way of doing this without having to write it out to disk.  `torch::jit::load()` can take an `istream`, so if you can just write it out to a buffer in memory that will work.

RaulPPelaez replied:
Turns out there was a way! hopefully it holds up for some versions :p

RaulPPelaez commented:
> ```
> E       CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
> E       For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
> ```
> 
> Like it says! If you set `CUDA_LAUNCH_BLOCKING=1`, you should get an error message that tells what the actual problem is.

Anticlimactically, I get the same exact message but without the CUDA_LAUNCH_BLOCKING hint -.- 
I believe this might be an issue with the latest nnpops conda package. This fails in my machine:

```shell
$ mamba create -n test
$ mamba activate test
$ mamba install nnpops==0.5 mdtraj pytest
$ cd path_to_NNPOPS_tests
$ pytest TestOptimizedTorchANI.py
```
All CUDA tests fail with the same error "CUDA error: no kernel image is available for execution on the device"

However, compiling the latest version of NNPops from source and running the same test yields no errors.

I realize now this was the reason I was not seeing the error at first, because I was using nnpops from source instead of conda.

This is the conda env I get from running the above:
<details>
  <summary>Click me</summary>
  
  ```shell
# packages in environment at /shared/raul/mambaforge/envs/test:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                  2_kmp_llvm    conda-forge
astunparse                1.6.3              pyhd8ed1ab_0    conda-forge
blosc                     1.21.3               hafa529b_0    conda-forge
brotlipy                  0.7.0           py310h5764c6d_1005    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.18.1               h7f98852_0    conda-forge
c-blosc2                  2.8.0                hf91038e_1    conda-forge
ca-certificates           2022.12.7            ha878542_0    conda-forge
certifi                   2022.12.7          pyhd8ed1ab_0    conda-forge
cffi                      1.15.1          py310h255011f_3    conda-forge
charset-normalizer        3.1.0              pyhd8ed1ab_0    conda-forge
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
cryptography              40.0.2          py310h34c0648_0    conda-forge
cuda-version              11.8                 h70ddcb2_1    conda-forge
cudatoolkit               11.8.0              h37601d7_11    conda-forge
cudnn                     8.8.0.121            h0800d71_0    conda-forge
exceptiongroup            1.1.1              pyhd8ed1ab_0    conda-forge
filelock                  3.12.0             pyhd8ed1ab_0    conda-forge
gmp                       6.2.1                h58526e2_0    conda-forge
gmpy2                     2.1.2           py310h3ec546c_1    conda-forge
hdf5                      1.14.0          nompi_hb72d44e_103    conda-forge
icu                       72.1                 hcb278e6_0    conda-forge
idna                      3.4                pyhd8ed1ab_0    conda-forge
importlib-metadata        6.6.0              pyha770c72_0    conda-forge
iniconfig                 2.0.0              pyhd8ed1ab_0    conda-forge
jinja2                    3.1.2              pyhd8ed1ab_1    conda-forge
keyutils                  1.6.1                h166bdaf_0    conda-forge
krb5                      1.20.1               h81ceb04_0    conda-forge
lark-parser               0.12.0             pyhd8ed1ab_0    conda-forge
ld_impl_linux-64          2.40                 h41732ed_0    conda-forge
libaec                    1.0.6                hcb278e6_1    conda-forge
libblas                   3.9.0           16_linux64_openblas    conda-forge
libcblas                  3.9.0           16_linux64_openblas    conda-forge
libcurl                   8.0.1                h588be90_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 12.2.0              h65d4601_19    conda-forge
libgfortran-ng            12.2.0              h69a702a_19    conda-forge
libgfortran5              12.2.0              h337968e_19    conda-forge
libhwloc                  2.9.1                hd6dc26d_0    conda-forge
libiconv                  1.17                 h166bdaf_0    conda-forge
liblapack                 3.9.0           16_linux64_openblas    conda-forge
libmagma                  2.7.1                hc72dce7_1    conda-forge
libmagma_sparse           2.7.1                hc72dce7_2    conda-forge
libnghttp2                1.52.0               h61bc06f_0    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libopenblas               0.3.21          pthreads_h78a6416_3    conda-forge
libprotobuf               3.21.12              h3eb15da_0    conda-forge
libsqlite                 3.40.0               h753d276_1    conda-forge
libssh2                   1.10.0               hf14f497_3    conda-forge
libstdcxx-ng              12.2.0              h46fd767_19    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libxml2                   2.10.4               hfdac1af_0    conda-forge
libzlib                   1.2.13               h166bdaf_4    conda-forge
llvm-openmp               16.0.2               h4dfa4b3_0    conda-forge
lz4-c                     1.9.4                hcb278e6_0    conda-forge
lzo                       2.10              h516909a_1000    conda-forge
magma                     2.7.1                ha770c72_2    conda-forge
markupsafe                2.1.2           py310h1fa729e_0    conda-forge
mdtraj                    1.9.7           py310h902c554_4    conda-forge
mkl                       2022.2.1         h84fe81f_16997    conda-forge
mpc                       1.3.1                hfe3b2da_0    conda-forge
mpfr                      4.2.0                hb012696_0    conda-forge
mpmath                    1.3.0              pyhd8ed1ab_0    conda-forge
nccl                      2.15.5.1             h0800d71_0    conda-forge
ncurses                   6.3                  h27087fc_1    conda-forge
networkx                  3.1                pyhd8ed1ab_0    conda-forge
nnpops                    0.5             cuda112py310h042f432_1    conda-forge
numexpr                   2.7.3           py310hb5077e9_1    conda-forge
numpy                     1.24.3          py310ha4c1d20_0    conda-forge
openssl                   3.1.0                hd590300_2    conda-forge
packaging                 23.1               pyhd8ed1ab_0    conda-forge
pandas                    2.0.1           py310h7cbd5c2_0    conda-forge
pip                       23.1.2             pyhd8ed1ab_0    conda-forge
platformdirs              3.4.0              pyhd8ed1ab_0    conda-forge
pluggy                    1.0.0              pyhd8ed1ab_5    conda-forge
pooch                     1.7.0              pyha770c72_3    conda-forge
py-cpuinfo                9.0.0              pyhd8ed1ab_0    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyopenssl                 23.1.1             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1              pyha2e5f31_6    conda-forge
pytables                  3.8.0           py310hde6a235_1    conda-forge
pytest                    7.3.1              pyhd8ed1ab_0    conda-forge
python                    3.10.10         he550d4f_0_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-tzdata             2023.3             pyhd8ed1ab_0    conda-forge
python_abi                3.10                    3_cp310    conda-forge
pytorch                   2.0.0           cuda112py310he33e0d6_200    conda-forge
pytz                      2023.3             pyhd8ed1ab_0    conda-forge
readline                  8.2                  h8228510_1    conda-forge
requests                  2.29.0             pyhd8ed1ab_0    conda-forge
scipy                     1.10.1          py310h8deb116_0    conda-forge
setuptools                67.7.2             pyhd8ed1ab_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
sleef                     3.5.1                h9b69904_2    conda-forge
snappy                    1.1.10               h9fff704_0    conda-forge
sympy                     1.11.1          pypyh9d50eac_103    conda-forge
tbb                       2021.9.0             hf52228f_0    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
tomli                     2.0.1              pyhd8ed1ab_0    conda-forge
torchani                  2.2                pyh9f0ad1d_0    conda-forge
typing-extensions         4.5.0                hd8ed1ab_0    conda-forge
typing_extensions         4.5.0              pyha770c72_0    conda-forge
tzdata                    2023c                h71feb2d_0    conda-forge
urllib3                   1.26.15            pyhd8ed1ab_0    conda-forge
wheel                     0.40.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
zipp                      3.15.0             pyhd8ed1ab_0    conda-forge
zlib                      1.2.13               h166bdaf_4    conda-forge
zlib-ng                   2.0.7                h0b41bf4_0    conda-forge
zstd                      1.5.2                h3eb15da_6    conda-forge
```

</details>

Notic how nnpops and pytorch are installed with cuda11.2, but cudatoolkit 11.8 is installed.

sef43 commented:
All my tests pass when using this PR. The only obvious difference in my conda env to yours is:
`torchani                  2.2.3           cuda112py310h73eb55c_0    conda-forge`

My system CUDA version is 12.1

peastman commented:
@RaulPPelaez what GPU do you have?

RaulPPelaez commented:
A 2080 Ti

peastman commented:
Here are a few pages relevant to that error:

https://github.com/pytorch/pytorch/issues/31285
https://stackoverflow.com/questions/65739700/pytorch-cuda-error-no-kernel-image-is-available-for-execution-on-the-device-on
https://forums.developer.nvidia.com/t/runtimeerror-cuda-error-no-kernel-image-is-available-for-execution-on-the-device/167708

Reading through them though, it isn't clear to me what the problem is.  Generally it seems to involve some incompatibility between your GPU, CUDA toolkit, and PyTorch version.  You have CUDA 11.8, which supports your GPU.  Your PyTorch package was built against CUDA 11.2, which is binary compatible with 11.8.  Is your driver new enough to support 11.8?

sef43 commented:
I will also try and check the NNPOps tests. I think I remember seeing this error before `"CUDA error: no kernel image is available for execution on the device"`, when using pytorch, however it was back when my system was NVIDIA driver 515, with CUDA 11.7. (When installing OpenMM I would have to force specify `cudatoolkit=11.7` in `conda install` for everything to work properly). I have recently updated to driver version 530 and CUDA 12.1 and have not seen the error since the upgrade.

mikemhenry commented:
@RaulPPelaez It _should_ be possible but on this page https://github.com/openmm/openmm-torch/actions/workflows/self-hosted-gpu-test.yml I don't see the workflow dispatch button (screenshot taken from another repo)
![image](https://user-images.githubusercontent.com/11765982/235255420-b4bfeffe-12a0-418f-855d-44b57c166d17.png)

And while that PR isn't merged yet, I did make it as a branch on the repository so I thought that button would show up. It would be really easy for me add the action to this pull request to test it, and then just delete the file so we can merge in this PR without pulling in the action.

Oh, we can't do that since you are working on a fork and the secrets for the action can only run in the main repo.

What I will do is make a new PR with the action + the changes in this one so we can see if it works.

RaulPPelaez commented:
I thought merging with your branch would run the gpu ci, then I understood your comments about the secrets -.-

mikemhenry commented:
@RaulPPelaez Sorry I didn't get to this before I went on vaction, but if you push this directly to the openmm-torch repo, then it should work if you merge in my branch

RaulPPelaez commented:
I have no push access to this repo I am afraid -.-

RaulPPelaez commented:
This is ready to merge on my part. The issues discussed above seem to be unrelated to openmm-torch, if any related to NNPOps, however I have tried now and all tests are passing.

peastman commented:
Looks good to me.

mikemhenry commented:
I'm back from my vacation and can run this branch in the CI.

mikemhenry commented:
Testing in this PR https://github.com/openmm/openmm-torch/pull/108

mikemhenry commented:
Hmmm sorry @RaulPPelaez I will fix this git mess, I was sure that I wasn't pushing to your branch but somewhere I messed up.

mikemhenry commented:
Ah I see what happened now, since we were working on these PRs at the same time, the merge (which this is expected behavior) merged in our commits so it is hard to pop off the changes without telling git which parent to use, I can fix this :)

RaulPPelaez commented:
Hi everyone, could we merge this so we can get a new conda release?
The issue that makes a test fail in the GPU ci still looks to me like an NNPops problem.
cc @mikemhenry
